### ISSUE TITLE
Gradient calculation of ODE solution really slow

### ISSUE THREAD

**Issue by vcantarella** [patrick-kidger/diffrax#226]
Hi I am trying to optimize a reactive transport model which is formulated as an ODE problem and solved with Diffrax.

The solution part of the model is this:

```
 from diffrax import ODETerm, Tsit5, diffeqsolve, PIDController, SaveAt
            
            term = ODETerm(self.model_build)

            solver = Tsit5()

            t0, t1, dt0 = stress_intervals[0], stress_intervals[-1], 30
            
            solution = diffeqsolve(term, solver, t0=t0, t1=t1, dt0=0.01, max_steps=20000,
                       stepsize_controller=PIDController(pcoeff=0.3, icoeff=0.3, dcoeff=0, rtol=1e-8, atol=1e-9),
                       saveat = SaveAt(ts=stress_intervals), 
                       y0=y0,
                       args = args,)
            
            #Running the ODE solver
            ode_output = solution.ys
```

Then I compute the loss function based on some data:

```
def loss_fun(model_params):
    res, x_steps, time_intervals = model_run(model_params)
    res = jnp.concatenate([h4 - res[idxs_hrs[0],idxs,4],
                     h8 - res[idxs_hrs[1],idxs,4],
                     h12 - res[idxs_hrs[2],idxs,4],
                     h16 - res[idxs_hrs[3],idxs,4],
                     h20 - res[idxs_hrs[4],idxs,4],
                     h30 - res[idxs_hrs[5],idxs,4],
                     h40 - res[idxs_hrs[-1],idxs,4],])
    least_squares = jnp.sum((res)**2)
    return least_squares
```

and calculate the gradient of this function:

`
grad(loss_function)(init_params)
`
which takes a long time and in the end I get a warining like this:

```
2023-02-14 09:17:51.924196: E external/org_tensorflow/tensorflow/compiler/xla/service/slow_operation_alarm.cc:133] The operation took 2m53.878393437s

********************************
[Compiling module jit_diffeqsolve] Very slow compile?  If you want to file a bug, run with envvar XLA_FLAGS=--xla_dump_to=/tmp/foo and attach the results.
********************************
2023-02-14 09:29:34.257165: E external/org_tensorflow/tensorflow/compiler/xla/service/slow_operation_alarm.cc:65] 
********************************
[Compiling module jit_diffeqsolve] Very slow compile?  If you want to file a bug, run with envvar XLA_FLAGS=--xla_dump_to=/tmp/foo and attach the results.
********************************
2023-02-14 09:30:06.501208: E external/org_tensorflow/tensorflow/compiler/xla/service/slow_operation_alarm.cc:133] The operation took 2m32.243956032s

********************************
[Compiling module jit_diffeqsolve] Very slow compile?  If you want to file a bug, run with envvar XLA_FLAGS=--xla_dump_to=/tmp/foo and attach the results.
********************************
```

Am I doing something wrong? Or can I do anything to better run gradient calculations for these types of situations?

Thank you!!

**Comment by patrick-kidger** [comment]
So a couple things do jump out. First, you have a very large value of `max_steps`. Until recently increasing this value would result in much slower runtime. However, I've just pushed v0.3.0 of Diffrax, which should fix this issue. Try upgrading and see if that helps?

Also, don't forget to wrap all of your JAX operations in a JIT (either `jax.jit` or `eqx.filter_jit`), as this is necessary for good performance.

How do things compare after you do this?
